hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsRegistryImpl.java

```diff
@@ -179,21 +179,17 @@ ProbeInstance getProbeInstance(String name) {
             return;
         }
 
-        synchronized (lockStripe.getLock(source)) {
-            ProbeInstance probeInstance = probeInstances.get(name);
-            if (probeInstance == null) {
-                probeInstance = new ProbeInstance<S>(name, source, function);
-                probeInstances.put(name, probeInstance);
-            } else {
-                logOverwrite(probeInstance);
-            }
-
-            if (logger.isFinestEnabled()) {
-                logger.finest("Registered probeInstance " + name);
+        ProbeInstance probeInstance = probeInstances.putIfAbsent(name, new ProbeInstance<S>(name, source, function));
+        if (probeInstance != null) {
+            logOverwrite(probeInstance);
+            synchronized (lockStripe.getLock(source)) {
+                probeInstance.source = source;
+                probeInstance.function = function;
             }
+        }
 
-            probeInstance.source = source;
-            probeInstance.function = function;
+        if (logger.isFinestEnabled()) {
+            logger.finest("Registered probeInstance " + name);
```

Comment:
Probably this may be logging falsely and `logOverwrite` too. If the object is taken from `putIfAbsent` gets removed concurrently, registration/overwrite doesn't happen. This behavior looks right, but the logging may be misleading. Checking `contains` in the critical section and logging only in that case might solve this.

Note: this code is inherently racy and we are about to fix it in 4.0 by changing explicit registration/deregistration to dynamic discovery.

Reply:
It that still planned for 4.0?

Reply:
@blazember You are right  I think, you might end up with a missing probe instance in this case, if the it's removed right after the putIfAbsent call. Should the probe be reinserted to the map instead of modifying the source/function? Is there a reason for modifying the existing registered one instead of replacing it?

Reply:
@cangencer We can add it back if we detect that the probe was removed since `putIfAbsent`, but in that case, we might leak. The replaced code may also leak or lose probe. I think there is no good solution with the code in 3.12.x. The PR proposes a solution that's in line with the replaced code. We use `putIfAbsent` to reason about the order of the execution of the concurrent threads, and if the order is `register-than-deregister`, we won't have the probe registered, just like with the replaced code where we serialized the threads on the lock. I would fix only the logging.

@pveentjer Yes, solving the lifecycle issues is a requirement of the metrics work for 4.0.